File: android/app/src/main/java/com/omniagent/assistant/agent/AgentOrchestrator.java
package com.omniagent.assistant.agent;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import com.omniagent.assistant.model.AgentDecision;
import com.omniagent.assistant.service.AgentAccessibilityService;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

public class AgentOrchestrator {

    public interface OrchestratorListener {
        void onLog(String message);
        void onFinished(boolean success, String summary);
    }

    private final Context context;
    private final LLMClient llmClient;
    private final OrchestratorListener listener;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final AtomicBoolean isRunning = new AtomicBoolean(false);

    public AgentOrchestrator(Context context, LLMClient llmClient, OrchestratorListener listener) {
        this.context = context;
        this.llmClient = llmClient;
        this.listener = listener;
    }

    private static class TaskRunner implements Runnable {
        private final AgentOrchestrator orchestrator;
        private final String goal;

        TaskRunner(AgentOrchestrator orchestrator, String goal) {
            this.orchestrator = orchestrator;
            this.goal = goal;
        }

        @Override
        public void run() {
            try {
                orchestrator.runLoop(goal);
            } catch (Exception e) {
                orchestrator.postLog("Subsystem anomaly: " + e.getMessage());
                orchestrator.postFinish(false, "Error: " + e.getMessage());
            } finally {
                orchestrator.isRunning.set(false);
            }
        }
    }

    public void start(String goal) {
        if (isRunning.getAndSet(true)) {
            stop();
            isRunning.set(true);
        }
        executor.execute(new TaskRunner(this, goal));
    }

    public void stop() {
        isRunning.set(false);
    }

    private void runLoop(String goal) {
        postLog("Jarvis Command Received: \"" + goal + "\"");

        // Step 1: Check for Direct Jarvis Subsystem Control (Flashlight, Volume, Apps, Alarms, Calls)
        JarvisBrain.ExecutionResult direct = JarvisBrain.tryDirectSystemControl(context, goal);
        if (direct.handled) {
            postLog("Status: " + direct.detail);
            postFinish(direct.success, direct.spokenFeedback);
            return;
        }

        // Step 2: Check Accessibility Engine for Universal App Interaction
        AgentAccessibilityService service = AgentAccessibilityService.instance;
        if (service == null) {
            postLog("Jarvis Accessibility Subsystem is not engaged. Please enable in Settings.");
            postFinish(false, "Accessibility service is required to control on-screen applications, sir.");
            return;
        }

        // Step 3: Dynamic Screen Execution
        postLog("Analyzing visual telemetry on active screen...");
        boolean dynamicHandled = JarvisBrain.executeDynamicScreenAction(service, goal);
        if (dynamicHandled) {
            postLog("Successfully executed action on active application.");
            postFinish(true, "Action executed successfully on screen, sir.");
            return;
        }

        // Step 4: Full Multi-Hop ReAct Loop (via Gemini / OpenAI or Semantic Planner)
        List<String> history = new ArrayList<String>();
        int step = 0;
        int maxSteps = 10;

        while (isRunning.get() && step < maxSteps) {
            step++;
            postLog("Step " + step + ": Inspecting UI hierarchy...");

            UIHierarchyParser.ParseResult screen = service.inspectScreen();
            AgentDecision decision = llmClient.decide(goal, screen.promptText, history);

            postLog("Jarvis Thought: " + decision.thought);
            postLog("Executing: " + decision.actionType + (decision.targetId > 0 ? " [Target: " + decision.targetId + "]" : ""));

            if ("finish".equalsIgnoreCase(decision.actionType)) {
                postFinish(decision.success, decision.message != null ? decision.message : "Task completed, sir.");
                return;
            }

            boolean ok = false;
            String act = decision.actionType.toLowerCase();
            if ("click".equals(act)) {
                if (decision.targetId > 0) {
                    ok = service.clickElement(decision.targetId);
                } else if (decision.coordX >= 0 && decision.coordY >= 0) {
                    ok = service.clickCoordinates(decision.coordX, decision.coordY);
                }
            } else if ("type".equals(act)) {
                ok = service.inputText(decision.targetId, decision.text != null ? decision.text : "");
            } else if ("scroll".equals(act)) {
                ok = service.scroll("down".equalsIgnoreCase(decision.direction));
            } else if ("launch_app".equals(act)) {
                ok = decision.packageName != null && service.launchApp(decision.packageName);
            }

            history.add("Step " + step + ": " + decision.actionType + " -> " + (ok ? "ok" : "failed"));

            try {
                Thread.sleep(1100);
            } catch (InterruptedException ignored) {}
        }

        if (step >= maxSteps) {
            postFinish(false, "Task sequence concluded, sir.");
        }
    }

    private static class LogPoster implements Runnable {
        private final OrchestratorListener listener;
        private final String message;
        LogPoster(OrchestratorListener listener, String message) {
            this.listener = listener;
            this.message = message;
        }
        @Override
        public void run() {
            if (listener != null) listener.onLog(message);
        }
    }

    private static class FinishPoster implements Runnable {
        private final OrchestratorListener listener;
        private final boolean success;
        private final String summary;
        FinishPoster(OrchestratorListener listener, boolean success, String summary) {
            this.listener = listener;
            this.success = success;
            this.summary = summary;
        }
        @Override
        public void run() {
            if (listener != null) listener.onFinished(success, summary);
        }
    }

    private void postLog(String msg) {
        mainHandler.post(new LogPoster(listener, msg));
    }

    private void postFinish(boolean success, String summary) {
        mainHandler.post(new FinishPoster(listener, success, summary));
    }
}

File: android/app/src/main/java/com/omniagent/assistant/agent/AgentOrchestrator.kt
package com.omniagent.assistant.agent

import android.content.Context
import android.util.Log
import com.omniagent.assistant.model.AgentDecision
import com.omniagent.assistant.model.GlobalKey
import com.omniagent.assistant.model.ScrollDirection
import com.omniagent.assistant.service.AgentAccessibilityService
import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Autonomous Agent Orchestrator.
 * Implements the Observe -> Reason -> Act -> Verify loop on Android.
 */
class AgentOrchestrator(
    private val context: Context,
    private val llmClient: LLMClient,
    private val listener: AgentListener? = null
) {

    interface AgentListener {
        fun onStateChanged(state: AgentState, detail: String = "")
        fun onActionExecuted(stepIndex: Int, decision: AgentDecision)
        fun onConfirmationRequired(question: String, onConfirm: (Boolean) -> Unit)
        fun onTaskFinished(success: Boolean, summary: String)
    }

    enum class AgentState {
        IDLE,
        THINKING,
        EXECUTING,
        AWAITING_CONFIRMATION,
        FINISHED,
        CANCELLED,
        ERROR
    }

    companion object {
        private const val TAG = "AgentOrchestrator"
        private const val MAX_STEPS = 15
    }

    private val isRunning = AtomicBoolean(false)
    private var agentJob: Job? = null
    private val actionHistory = mutableListOf<String>()

    val isActive: Boolean
        get() = isRunning.get()

    fun startTask(voiceGoal: String) {
        if (isRunning.getAndSet(true)) {
            Log.w(TAG, "Task already running, aborting previous.")
            stopTask()
            isRunning.set(true)
        }

        actionHistory.clear()
        agentJob = CoroutineScope(Dispatchers.Default).launch {
            try {
                runAgentLoop(voiceGoal)
            } catch (e: CancellationException) {
                Log.i(TAG, "Agent task cancelled by user.")
                withContext(Dispatchers.Main) {
                    listener?.onStateChanged(AgentState.CANCELLED, "Cancelled by user")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error in agent loop", e)
                withContext(Dispatchers.Main) {
                    listener?.onStateChanged(AgentState.ERROR, e.localizedMessage ?: "Unknown error")
                }
            } finally {
                isRunning.set(false)
            }
        }
    }

    fun stopTask() {
        isRunning.set(false)
        agentJob?.cancel()
        agentJob = null
    }

    private suspend fun runAgentLoop(goal: String) {
        val service = AgentAccessibilityService.instance
        if (service == null) {
            withContext(Dispatchers.Main) {
                listener?.onStateChanged(
                    AgentState.ERROR,
                    "Accessibility Service is not enabled. Please enable it in Settings."
                )
            }
            return
        }

        var step = 0
        while (isRunning.get() && step < MAX_STEPS) {
            step++
            withContext(Dispatchers.Main) {
                listener?.onStateChanged(AgentState.THINKING, "Step $step: Analyzing screen...")
            }

            // 1. Observe
            val screenState = service.inspectCurrentScreen()

            // 2. Reason
            val decision = llmClient.decideNextAction(
                goal = goal,
                screenHierarchy = screenState.promptRepresentation,
                actionHistory = actionHistory
            )

            Log.i(TAG, "Step $step Decision: thought=${decision.thought}, action=${decision.actionType}")

            // 3. Handle Sensitive Confirmation
            if (decision.actionType.equals("ask_confirmation", ignoreCase = true)) {
                withContext(Dispatchers.Main) {
                    listener?.onStateChanged(AgentState.AWAITING_CONFIRMATION, decision.message ?: "Confirmation required")
                }
                val confirmed = awaitUserConfirmation(decision.message ?: "Authorize action?")
                if (!confirmed) {
                    withContext(Dispatchers.Main) {
                        listener?.onTaskFinished(false, "User declined sensitive action.")
                    }
                    return
                }
                actionHistory.add("User confirmed: ${decision.message}")
                continue
            }

            // 4. Handle Finish
            if (decision.actionType.equals("finish", ignoreCase = true)) {
                withContext(Dispatchers.Main) {
                    listener?.onActionExecuted(step, decision)
                    listener?.onTaskFinished(decision.success, decision.message ?: "Task finished.")
                }
                return
            }

            // 5. Act
            withContext(Dispatchers.Main) {
                listener?.onStateChanged(AgentState.EXECUTING, decision.thought)
                listener?.onActionExecuted(step, decision)
            }

            val success = executeDecision(service, decision)
            val historyEntry = "Action: ${decision.actionType} (target=${decision.targetId ?: decision.coordinates}), result=${if (success) "success" else "failed"}"
            actionHistory.add(historyEntry)

            // 6. Settle delay
            delay(1000L)
        }

        if (step >= MAX_STEPS) {
            withContext(Dispatchers.Main) {
                listener?.onTaskFinished(false, "Reached maximum step limit ($MAX_STEPS).")
            }
        }
    }

    private suspend fun executeDecision(
        service: AgentAccessibilityService,
        decision: AgentDecision
    ): Boolean {
        return when (decision.actionType.lowercase()) {
            "click" -> {
                when {
                    decision.targetId != null -> service.clickElement(decision.targetId)
                    decision.coordinates != null && decision.coordinates.size >= 2 ->
                        service.clickCoordinate(decision.coordinates[0], decision.coordinates[1])
                    else -> false
                }
            }
            "type" -> {
                val text = decision.text ?: ""
                service.inputText(decision.targetId, text)
            }
            "scroll" -> {
                val dir = when (decision.direction?.lowercase()) {
                    "up" -> ScrollDirection.UP
                    "left" -> ScrollDirection.LEFT
                    "right" -> ScrollDirection.RIGHT
                    else -> ScrollDirection.DOWN
                }
                service.scroll(dir, decision.targetId)
            }
            "swipe" -> {
                val coords = decision.coordinates ?: listOf(500f, 1500f, 500f, 500f)
                if (coords.size >= 4) {
                    service.swipe(coords[0], coords[1], coords[2], coords[3])
                } else {
                    false
                }
            }
            "launch_app" -> {
                decision.packageName?.let { service.launchApp(it) } ?: false
            }
            "press_key" -> {
                val key = when (decision.key?.lowercase()) {
                    "back" -> GlobalKey.BACK
                    "home" -> GlobalKey.HOME
                    "recents" -> GlobalKey.RECENTS
                    "notifications" -> GlobalKey.NOTIFICATIONS
                    else -> GlobalKey.BACK
                }
                service.performGlobal(key)
            }
            "wait" -> {
                delay(1200L)
                true
            }
            else -> {
                Log.w(TAG, "Unknown action: ${decision.actionType}")
                false
            }
        }
    }

    private suspend fun awaitUserConfirmation(prompt: String): Boolean = suspendCancellableCoroutine { cont ->
        listener?.onConfirmationRequired(prompt) { approved ->
            if (cont.isActive) {
                cont.resume(approved) {}
            }
        }
    }
}
